model_compression_toolkit/gptq/keras/quantizer/soft_rounding/symmetric_soft_quantizer.py

```diff
@@ -0,0 +1,347 @@
+            threshold_values)
+        self.k_threshold = len(self.threshold_values) if self.per_axis else 1
+
+        # gamma and zeta are stretch parameters for computing the rectified sigmoind function.
+        # beta is used to set the regularization term.
+        # See: https://arxiv.org/pdf/2004.10568.pdf
+        self.gamma = SOFT_ROUNDING_GAMMA
+        self.zeta = SOFT_ROUNDING_ZETA
+        self.beta = SOFT_ROUNDING_BETA
+
+        self.quantizer_parameters = {}
+
+        # Initializing the temperature decay according to the number of expected gradient steps
+        if n_batches is None:
+            Logger.warning(f"Number of batches is not set correctly for the Soft Quantizer. A default value of "  # pragma: no cover
+                           f"{MAX_ITERATIONS_DEFAULT} is used to set the temperature decay which may affect the results.")
+
+        init_decay = MAX_ITERATIONS_DEFAULT if n_batches is None else n_epochs * n_batches
+        self.linear_decay = LinearTempDecay(init_decay)
+
+    def build(self,
+              tensor_shape: TensorShape,
+              name: str,
+              layer: QuantizeWrapper) -> Dict[str, tf.Variable]:
+        """
+        Add variables to the quantizer.
+
+        Args:
+            tensor_shape: Tensor shape the quantizer quantize.
+            name: Prefix of variables names.
+            layer: Layer to add the variables to. The variables are saved
+            in the layer's scope.
+
+        Returns:
+            Dictionary of new variables.
+        """
+
+        super().build(tensor_shape, name, layer)
+
+        if self.per_axis:
+            input_shape = tensor_shape
+            n_axis = len(input_shape)
+            quantization_axis = n_axis + self.quantization_axis if self.quantization_axis < 0 else \
+                self.quantization_axis
+            reshape_shape = [self.k_threshold if i == quantization_axis else 1 for i in range(n_axis)]
+        else:
+            reshape_shape = [self.k_threshold]
+
+        ar_iter = layer.add_weight(
+            name + gptq_constants.GPTQ_ITER,
+            shape=(),
+            initializer=tf.keras.initializers.Constant(0.0),
+            trainable=False)
+
+        ptq_threshold_tensor = layer.add_weight(
+            name + PTQ_THRESHOLD,
+            shape=reshape_shape,
+            initializer=tf.keras.initializers.Constant(1.0),
+            trainable=False)
+        ptq_threshold_tensor.assign(self.threshold_values.reshape(reshape_shape))
+
+        w = getattr(layer.layer, name)
+        auxvar_tensor = layer.add_weight(
+            name + gptq_constants.AUXVAR,
+            shape=[*w.shape],
+            initializer=tf.keras.initializers.Constant(0.0),
+            trainable=True)
+
+        delta = qutils.calculate_delta(ptq_threshold_tensor, self.num_bits, self.signed)
+        w_floor = tf.floor(w / delta)
```

Comment:
what if delta is 0?

Reply:
We divide by delta like this in all of our quantizers (not only in GPTQ).
I'm not sure that adding epsilon would not affect the quantization in some cases.
I prefer to leave it like this for now (as in the rest of the code).